File: src/game/sim/hex.ts
export type Axial = { q: number; r: number };

export const HEX_DIRS: Axial[] = [
  { q: 1, r: 0 },
  { q: 1, r: -1 },
  { q: 0, r: -1 },
  { q: -1, r: 0 },
  { q: -1, r: 1 },
  { q: 0, r: 1 },
];

export function hexKey(h: Axial): string {
  return `${h.q},${h.r}`;
}

export function hexEq(a: Axial, b: Axial): boolean {
  return a.q === b.q && a.r === b.r;
}

export function hexAdd(a: Axial, b: Axial): Axial {
  return { q: a.q + b.q, r: a.r + b.r };
}

export function hexSub(a: Axial, b: Axial): Axial {
  return { q: a.q - b.q, r: a.r - b.r };
}

export function hexScale(a: Axial, n: number): Axial {
  return { q: a.q * n, r: a.r * n };
}

export function rotate60(h: Axial, times: number): Axial {
  let q = h.q;
  let r = h.r;
  const t = ((times % 6) + 6) % 6;
  for (let i = 0; i < t; i++) {
    const nq = -r;
    const nr = q + r;
    q = nq;
    r = nr;
  }
  return { q, r };
}

export function cubeRound(qf: number, rf: number): Axial {
  const sf = -qf - rf;
  let q = Math.round(qf);
  let r = Math.round(rf);
  let s = Math.round(sf);
  const qd = Math.abs(q - qf);
  const rd = Math.abs(r - rf);
  const sd = Math.abs(s - sf);
  if (qd > rd && qd > sd) q = -r - s;
  else if (rd > sd) r = -q - s;
  return { q, r };
}

export function axialToPixel(h: Axial, size: number): { x: number; y: number } {
  return {
    x: size * Math.sqrt(3) * (h.q + h.r / 2),
    y: size * (3 / 2) * h.r,
  };
}

export function pixelToAxial(x: number, y: number, size: number): Axial {
  const q = ((Math.sqrt(3) / 3) * x - (1 / 3) * y) / size;
  const r = ((2 / 3) * y) / size;
  return cubeRound(q, r);
}

export function hexDistance(a: Axial, b: Axial): number {
  return (
    (Math.abs(a.q - b.q) +
      Math.abs(a.q + a.r - b.q - b.r) +
      Math.abs(a.r - b.r)) /
    2
  );
}

export function hexNeighbors(h: Axial): Axial[] {
  return HEX_DIRS.map((d) => hexAdd(h, d));
}

export function hexDisc(center: Axial, radius: number): Axial[] {
  const out: Axial[] = [];
  for (let q = -radius; q <= radius; q++) {
    const rMin = Math.max(-radius, -q - radius);
    const rMax = Math.min(radius, -q + radius);
    for (let r = rMin; r <= rMax; r++) {
      out.push({ q: center.q + q, r: center.r + r });
    }
  }
  return out;
}

export function hexRing(center: Axial, radius: number): Axial[] {
  if (radius <= 0) return [{ ...center }];
  return hexDisc(center, radius).filter((h) => hexDistance(center, h) === radius);
}

export function hexLine(a: Axial, b: Axial): Axial[] {
  const n = hexDistance(a, b);
  if (n === 0) return [{ ...a }];
  const out: Axial[] = [];
  for (let i = 0; i <= n; i++) {
    const t = i / n;
    out.push(cubeRound(a.q + (b.q - a.q) * t, a.r + (b.r - a.r) * t));
  }
  return out;
}

export function hexCone(origin: Axial, facing: number, range: number): Axial[] {
  const fwd = HEX_DIRS[((facing % 6) + 6) % 6];
  const fq = fwd.q;
  const fr = fwd.r;
  const fs = -fq - fr;
  const out: Axial[] = [];
  for (const h of hexDisc(origin, range)) {
    const d = hexDistance(origin, h);
    if (d === 0) continue;
    const dq = h.q - origin.q;
    const dr = h.r - origin.r;
    const ds = -dq - dr;
    const dot = fq * dq + fr * dr + fs * ds;
    if (dot >= d - 0.5) out.push(h);
  }
  return out;
}

export function hexSweep(origin: Axial, facing: number): Axial[] {
  const f = ((facing % 6) + 6) % 6;
  return [HEX_DIRS[(f + 5) % 6], HEX_DIRS[f], HEX_DIRS[(f + 1) % 6]].map((d) =>
    hexAdd(origin, d),
  );
}

export function facingToward(from: Axial, to: Axial): number {
  const dq = to.q - from.q;
  const dr = to.r - from.r;
  if (dq === 0 && dr === 0) return 0;
  let best = 0;
  let bestDot = -Infinity;
  HEX_DIRS.forEach((d, i) => {
    const ds = -d.q - d.r;
    const hs = -dq - dr;
    const dot = d.q * dq + d.r * dr + ds * hs;
    if (dot > bestDot) {
      bestDot = dot;
      best = i;
    }
  });
  return best;
}

export function placeFootprint(
  shape: Axial[],
  origin: Axial,
  facing: number,
): Axial[] {
  return shape.map((h) => hexAdd(origin, rotate60(h, facing)));
}

export function hexCorners(cx: number, cy: number, size: number): { x: number; y: number }[] {
  const out: { x: number; y: number }[] = [];
  for (let i = 0; i < 6; i++) {
    const angle = (Math.PI / 180) * (60 * i - 30);
    out.push({ x: cx + size * Math.cos(angle), y: cy + size * Math.sin(angle) });
  }
  return out;
}

File: src/hex.rs
//! Pointy-top axial hexes.

#[derive(
    Clone, Copy, Debug, Default, PartialEq, Eq, Hash, serde::Serialize, serde::Deserialize,
)]
pub struct Axial {
    pub q: i32,
    pub r: i32,
}

pub const HEX_DIRS: [Axial; 6] = [
    Axial { q: 1, r: 0 },
    Axial { q: 1, r: -1 },
    Axial { q: 0, r: -1 },
    Axial { q: -1, r: 0 },
    Axial { q: -1, r: 1 },
    Axial { q: 0, r: 1 },
];

impl Axial {
    pub const fn new(q: i32, r: i32) -> Self {
        Self { q, r }
    }

    pub fn key(self) -> (i32, i32) {
        (self.q, self.r)
    }

    pub fn add(self, o: Self) -> Self {
        Self {
            q: self.q + o.q,
            r: self.r + o.r,
        }
    }

    pub fn sub(self, o: Self) -> Self {
        Self {
            q: self.q - o.q,
            r: self.r - o.r,
        }
    }
}

pub fn hex_eq(a: Axial, b: Axial) -> bool {
    a.q == b.q && a.r == b.r
}

pub fn rotate60(h: Axial, times: i32) -> Axial {
    let mut q = h.q;
    let mut r = h.r;
    let t = ((times % 6) + 6) % 6;
    for _ in 0..t {
        let nq = -r;
        let nr = q + r;
        q = nq;
        r = nr;
    }
    Axial { q, r }
}

pub fn cube_round(qf: f32, rf: f32) -> Axial {
    let sf = -qf - rf;
    let mut q = qf.round();
    let mut r = rf.round();
    let s = sf.round();
    let qd = (q - qf).abs();
    let rd = (r - rf).abs();
    let sd = (s - sf).abs();
    if qd > rd && qd > sd {
        q = -r - s;
    } else if rd > sd {
        r = -q - s;
    }
    Axial {
        q: q as i32,
        r: r as i32,
    }
}

pub fn axial_to_world(h: Axial, size: f32) -> (f32, f32) {
    let x = size * 3f32.sqrt() * (h.q as f32 + h.r as f32 / 2.0);
    let z = size * 1.5 * h.r as f32;
    (x, z)
}

pub fn world_to_axial(x: f32, z: f32, size: f32) -> Axial {
    let q = ((3f32.sqrt() / 3.0) * x - (1.0 / 3.0) * z) / size;
    let r = ((2.0 / 3.0) * z) / size;
    cube_round(q, r)
}

pub fn hex_distance(a: Axial, b: Axial) -> i32 {
    (i32::abs(a.q - b.q) + i32::abs(a.q + a.r - b.q - b.r) + i32::abs(a.r - b.r)) / 2
}

pub fn hex_neighbors(h: Axial) -> [Axial; 6] {
    HEX_DIRS.map(|d| h.add(d))
}

pub fn hex_disc(center: Axial, radius: i32) -> Vec<Axial> {
    let mut out = Vec::new();
    for q in -radius..=radius {
        let r_min = i32::max(-radius, -q - radius);
        let r_max = i32::min(radius, -q + radius);
        for r in r_min..=r_max {
            out.push(Axial {
                q: center.q + q,
                r: center.r + r,
            });
        }
    }
    out
}

pub fn hex_ring(center: Axial, radius: i32) -> Vec<Axial> {
    if radius <= 0 {
        return vec![center];
    }
    hex_disc(center, radius)
        .into_iter()
        .filter(|h| hex_distance(center, *h) == radius)
        .collect()
}

pub fn hex_line(a: Axial, b: Axial) -> Vec<Axial> {
    let n = hex_distance(a, b);
    if n == 0 {
        return vec![a];
    }
    (0..=n)
        .map(|i| {
            let t = i as f32 / n as f32;
            cube_round(
                a.q as f32 + (b.q - a.q) as f32 * t,
                a.r as f32 + (b.r - a.r) as f32 * t,
            )
        })
        .collect()
}

pub fn hex_cone(origin: Axial, facing: i32, range: i32) -> Vec<Axial> {
    let fwd = HEX_DIRS[(((facing % 6) + 6) % 6) as usize];
    let fq = fwd.q;
    let fr = fwd.r;
    let fs = -fq - fr;
    hex_disc(origin, range)
        .into_iter()
        .filter(|h| {
            let d = hex_distance(origin, *h);
            if d == 0 {
                return false;
            }
            let dq = h.q - origin.q;
            let dr = h.r - origin.r;
            let ds = -dq - dr;
            let dot = fq * dq + fr * dr + fs * ds;
            dot as f32 >= d as f32 - 0.5
        })
        .collect()
}

pub fn hex_sweep(origin: Axial, facing: i32) -> Vec<Axial> {
    let f = ((facing % 6) + 6) % 6;
    let dirs = [
        HEX_DIRS[((f + 5) % 6) as usize],
        HEX_DIRS[f as usize],
        HEX_DIRS[((f + 1) % 6) as usize],
    ];
    dirs.into_iter().map(|d| origin.add(d)).collect()
}

pub fn facing_toward(from: Axial, to: Axial) -> i32 {
    let dq = to.q - from.q;
    let dr = to.r - from.r;
    if dq == 0 && dr == 0 {
        return 0;
    }
    let mut best = 0;
    let mut best_dot = i32::MIN;
    for (i, d) in HEX_DIRS.iter().enumerate() {
        let ds = -d.q - d.r;
        let hs = -dq - dr;
        let dot = d.q * dq + d.r * dr + ds * hs;
        if dot > best_dot {
            best_dot = dot;
            best = i as i32;
        }
    }
    best
}

pub fn place_footprint(shape: &[Axial], origin: Axial, facing: i32) -> Vec<Axial> {
    shape
        .iter()
        .map(|h| origin.add(rotate60(*h, facing)))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn distance_neighbors() {
        let o = Axial::new(3, 4);
        for n in hex_neighbors(o) {
            assert_eq!(hex_distance(o, n), 1);
        }
        assert_eq!(hex_distance(o, o), 0);
    }

    #[test]
    fn world_roundtrip() {
        let h = Axial::new(4, 2);
        let (x, z) = axial_to_world(h, 1.0);
        assert_eq!(world_to_axial(x, z, 1.0), h);
    }

    #[test]
    fn rotate_full_turn() {
        let h = Axial::new(2, -1);
        assert_eq!(rotate60(h, 6), h);
    }

    #[test]
    fn sweep_is_three() {
        assert_eq!(hex_sweep(Axial::new(0, 0), 0).len(), 3);
    }
}
